# Resolucion de ecuaciones difeenciales con sympy [sympy]

<b>Deadline:</b> Tuesday, 24 de May de 2022, 00:00

Dada la siguiente representación alternativa de la ecuación cuadrática del sistema poblacional, resolverla utilizando el módulo sympy de Python3:

$$
\frac{\mathrm{d} x(t)}{\mathrm{d} x} = \alpha x_n + \beta x_n^2
$$


Entregar un programa Python que:
1. Resuelva la ecuación paso a paso y muestre el resultado final por pantalla.
2. Tenga documentadas las funciones creadas y la secuencia de pasos hasta alcanzar el resultado.
3. Explicación breve qué representa la ecuación, y una breve introducción al código que resuelve el problema.


<b>Requisitos de la entrega</b>:
La entrega deberá cumplir los siguientes requisitos:
- El código deberá ser presentado en un repositorio git
- Se creará un tag/release para la versión "estable" del programa entregable. El tag deberá ser creado antes de la fecha límite de presentación.

Además del código propiamente dicho, y convenientemente ordenado, deberá contar con los siguientes archivos en el repositorio GIT:

- <b>INSTALL.md</b>: contendrá las instrucciones para clonar e instalar/lanzar la aplicación.
- <b>README.md</b>: contendrá la ayuda y uso básico de la aplicación.
- <b>README.info</b>: contendrá un informe sobre las decisiones principales de diseñodel sistema, y su justificación (por qué usar tal o cual método, paquete, módulo, etc, y cómo se implementaron las reglas del juego de la vida.
- <b>README.todo</b>: contendrá una lista detallada de mejoras características/features del sistema para futuras correcciones y versiones.

In [21]:
import sympy as sp

Primero armamos el lado izquierdo de la ecuacion diferencial

In [22]:
t = sp.symbols("t")
x = sp.Function("x")

dxdt = sp.diff(x(t),t)
dxdt

Derivative(x(t), t)

Definimos las variables alfa y beta del modelo cuadrático, para armar la parte derecha de la ecuacion.

Después, relacionamos todos los elementos para armar la ecuacion a calcular su derivada de primer grado

In [23]:
alpha = sp.symbols(r"\alpha")
beta = sp.symbols(r"\beta")

ecuacion_diferencial = sp.Eq(dxdt, alpha * x(t) + beta * x(t) ** 2)
ecuacion_diferencial

Eq(Derivative(x(t), t), \alpha*x(t) + \beta*x(t)**2)

Obtenemos el valor de la derivada de primer grado

In [24]:
derivada_prim_grado = sp.dsolve(ecuacion_diferencial)
derivada_prim_grado

Eq(x(t), \alpha*exp(\alpha*(C1 + t))/(\beta*(1 - exp(\alpha*(C1 + t)))))

Queremos evaluar a la ecuacion cuando $t=0$, por lo tanto, guardamos la parte derecha de la ecuacion y la evaluamos en el tiempo inicial.

In [25]:
parte_derecha = derivada_prim_grado.rhs
at0 = parte_derecha.subs(t, 0)
at0

\alpha*exp(C1*\alpha)/(\beta*(1 - exp(C1*\alpha)))

Entonces, este valor obtenido nos el valor de la poblacion inicial $x_0$ ya que estamos evaluando la ecuacion en $t_0$.

Creamos entonces los simbolos $C_1$ y $x_0$ para poder plantear la ecuacion algebraica que nos servirá para reemplazar $C_1$ en la parte derecha de la ecuacion general.

In [26]:
C1, x_0 = sp.symbols("C1 x_0")

ec_algebraica = sp.Eq(at0, x_0)
ec_algebraica

Eq(\alpha*exp(C1*\alpha)/(\beta*(1 - exp(C1*\alpha))), x_0)

Resolvemos la ecuacion algebraica

In [27]:
solucionC1 = sp.solve(ec_algebraica, C1)[0]
solucionC1

log(\beta*x_0/(\alpha + \beta*x_0))/\alpha

Reemplazamos la solucion obtenida en la parte derecha de la ecuacion general, para obtener la solución particular.

In [28]:
sol_particular = parte_derecha.subs(C1, solucionC1)
sol_particular = sp.simplify(sol_particular)
sol_particular

\alpha*x_0*exp(\alpha*t)/(\alpha - \beta*x_0*exp(\alpha*t) + \beta*x_0)

Como podemos ver, si reemplazamos el valor de $t=0$, obtenemos la poblacion inicial $x_0$

In [29]:
sol_particular.subs(t, 0)

x_0

Si queremos analizar el desempeño de la funcion en valores especificos, podemos compararla con los valores historicos de EE.UU y de las Naciones Unidas. Tengamos en cuenta que el año donde se empieza a tener un registro poblacional es el año 1950, por lo que éste sera nuestro $t_0$

1. Compararemos los resultados del año 2010
2. Compararemos los resultados del año 1980

In [30]:
valores = [(alpha, 0.025), (beta, -0.0018), (x_0, 2.56), (t, 60)]
en_t60 = sol_particular.subs(valores)
en_t60

6.98837183188134

In [31]:
valores = [(alpha, 0.025), (beta, -0.0018), (x_0, 2.56), (t, 30)]
en_t30 = sol_particular.subs(valores)
en_t30

4.49422461620145

**Conclusiones:**
- Para el año **2010**, las poblaciones que se obtuvieron fueron las siguientes (en mil millones): EEUU: 6,858, NU: 6,916, Derivada de ec cuadratica: 6,988
- Para el año **1980**, las poblaciones que se obtuvieron fueron las siguientes (en mil millones): EEUU: 4,451, NU: 4,449, Derivada de ec cuadratica: 4.494

Podemos concluir entonces, que este modelo no está tan alejado de la realidad comparando los datos obtenidos.

In [6]:
C1, x_0 = sp.symbols("C1 x_0")

ec_algebraica = sp.Eq(at0, x_0)
ec_algebraica

Eq(\alpha*exp(C1*\alpha)/(\beta*(1 - exp(C1*\alpha))), x_0)

Resolvemos la ecuacion algebraica

In [7]:
solucionC1 = sp.solve(ec_algebraica, C1)[0]
solucionC1

log(\beta*x_0/(\alpha + \beta*x_0))/\alpha

Reemplazamos la solucion obtenida en la parte derecha de la ecuacion general, para obtener la solución particular.

In [8]:
sol_particular = parte_derecha.subs(C1, solucionC1)
sol_particular = sp.simplify(sol_particular)
sol_particular

\alpha*x_0*exp(\alpha*t)/(\alpha - \beta*x_0*exp(\alpha*t) + \beta*x_0)

Como podemos ver, si reemplazamos el valor de $t=0$, obtenemos la poblacion inicial $x_0$

In [9]:
sol_particular.subs(t, 0)

x_0

Si queremos analizar el desempeño de la funcion en valores especificos, podemos compararla con los valores historicos de EE.UU y de las Naciones Unidas. Tengamos en cuenta que el año donde se empieza a tener un registro poblacional es el año 1950, por lo que éste sera nuestro $t_0$

1. Compararemos los resultados del año 2010
2. Compararemos los resultados del año 1980

In [18]:
valores = [(alpha, 0.025), (beta, -0.0018), (x_0, 2.56), (t, 60)]
en_t60 = sol_particular.subs(valores)
en_t60

6.98837183188134

In [19]:
valores = [(alpha, 0.025), (beta, -0.0018), (x_0, 2.56), (t, 30)]
en_t30 = sol_particular.subs(valores)
en_t30

4.49422461620145

**Conclusiones:**
- Para el año **2010**, las poblaciones que se obtuvieron fueron las siguientes (en mil millones): EEUU: 6,858, NU: 6,916, Derivada de ec cuadratica: 6,988
- Para el año **1980**, las poblaciones que se obtuvieron fueron las siguientes (en mil millones): EEUU: 4,451, NU: 4,449, Derivada de ec cuadratica: 4.494

Podemos concluir entonces, que este modelo no está tan alejado de la realidad comparando los datos obtenidos.